product_default_image/models/product_template.py

```diff
@@ -0,0 +1,299 @@
+                    down to medium and small values.
+
+                    Note that to_type is required even if to_img_bg
+                    arg is supplied.
+
+                in_cache:
+                    Set True if calling this method to change the images
+                    in cache (helpful when calling from an onchange method
+                    in product.template)
+
+            Returns:
+                recordset if tmpl images have been changed.
+
+                None/False if any of below conditions are True.
+
+        """
+        if any([
+                to_type not in [
+                    'default_global', 'default_category',
+                    'no_image', 'custom'],
+                to_type == 'custom' and not to_img_bg,
+                to_type == 'no_image' and to_img_bg,
+                not isinstance(in_cache, bool)]):
+            return
+
+        # {img_bg: records}
+        write_map = defaultdict(lambda: self.env['product.template'].browse())
+
+        if to_img_bg:
+            to_img_bg = tools.image_resize_image_big(to_img_bg)
+            write_map[to_img_bg] += self
+
+        elif to_type == 'no_image':
+            write_map[None] += self
+
+        elif to_type == 'default_global':
+            company = self.env.user.company_id
+            write_map[company.product_image] += self
+
+        elif to_type == 'default_category':
+            for record in self:
+                write_map[record.categ_id.image] += record
+
+        if in_cache:
+
+            for img_bg, records in write_map.iteritems():
+
+                img_md = tools.image_resize_image_medium(img_bg)
+                img_sm = tools.image_resize_image_small(img_bg)
+
+                for record in records:
+                    record.image = img_bg
+                    record.image_medium = img_md
+                    record.image_small = img_sm
+                    record.img_type = to_type
+
+        else:
+
+            for img_bg, records in write_map.iteritems():
+
+                for record in records:
+                    record.write({
+                        'image': img_bg,
+                        'img_type': to_type,
+                    })
+                # records.write({
+                #     'image': img_bg,
+                #     'img_type': to_type,
+                # })
+
+        return self
+
+    @api.multi
+    def apply_default_image(self):
+
+        self.ensure_one()
+        target = self.env.user.company_id.product_image_target
+        self.auto_change_image = True
+
+        to_keys = {
+            'none': 'no_image',
+            'global': 'default_global',
+            'category': 'default_category',
+            'global_category': 'default_category',
+        }
+
+        if target not in to_keys:
+            return
+
+        if target == 'global_category':
+            categ = self.categ_id
+            if not any([categ.image, categ.image_medium, categ.image_small]):
```

Comment:
Should use tuples instead of arrays in these `any`s. Tuples are significantly lower overhead because they're immutable. We don't want mutable, so that's fine here.